File: android/app/src/main/java/com/antnagi/nagisheart/data/GameState.kt
package com.antnagi.nagisheart.data

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.int
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonPrimitive

class GameState(variablesData: VariablesData) {

    private val values = mutableMapOf<String, Any>()

    init {
        for ((name, def) in variablesData.numeric) {
            values[name] = def.initial ?: def.default ?: 0
        }
        for ((name, def) in variablesData.flags) {
            values[name] = resolveInitial(def)
        }
    }

    private fun resolveInitial(def: FlagVarDef): Any {
        val init = def.initial ?: return when (def.type) {
            "boolean" -> false
            else -> ""
        }
        if (init is JsonNull) return when (def.type) {
            "boolean" -> false
            else -> ""
        }
        val prim = init.jsonPrimitive
        return when {
            prim.booleanOrNull != null -> prim.boolean
            prim.intOrNull != null -> prim.int
            else -> prim.content
        }
    }

    fun get(name: String): Any? = values[name]

    fun getInt(name: String): Int = (values[name] as? Number)?.toInt() ?: 0

    fun getString(name: String): String = values[name]?.toString() ?: ""

    fun getBoolean(name: String): Boolean = values[name] as? Boolean ?: false

    fun set(name: String, value: Any) {
        values[name] = value
    }

    fun applyEffect(effect: Effect) {
        val prim = effect.`val`.jsonPrimitive
        when (effect.op) {
            "add" -> {
                val current = getInt(effect.`var`)
                values[effect.`var`] = current + prim.int
            }
            "set" -> {
                values[effect.`var`] = when {
                    prim.booleanOrNull != null -> prim.boolean
                    prim.intOrNull != null -> prim.int
                    else -> prim.content
                }
            }
        }
    }

    fun applyEffects(effects: List<Effect>) {
        effects.forEach { applyEffect(it) }
    }

    fun snapshot(): Map<String, Any> = values.toMap()

    fun toSerializable(): Map<String, JsonElement> {
        return values.mapValues { (_, v) ->
            when (v) {
                is Boolean -> JsonPrimitive(v)
                is Int -> JsonPrimitive(v)
                is Long -> JsonPrimitive(v)
                is Float -> JsonPrimitive(v)
                is Double -> JsonPrimitive(v)
                is String -> JsonPrimitive(v)
                else -> JsonPrimitive(v.toString())
            }
        }
    }

    fun restoreFrom(saved: Map<String, JsonElement>) {
        values.clear()
        for ((k, v) in saved) {
            val prim = v.jsonPrimitive
            values[k] = when {
                prim.booleanOrNull != null -> prim.boolean
                prim.intOrNull != null -> prim.int
                else -> prim.content
            }
        }
    }
}

File: web/src/engine/GameState.js
export class GameState {
  constructor(variablesData) {
    this._values = new Map();

    if (variablesData.numeric) {
      for (const [name, def] of Object.entries(variablesData.numeric)) {
        this._values.set(name, def.initial ?? def.default ?? 0);
      }
    }
    if (variablesData.flags) {
      for (const [name, def] of Object.entries(variablesData.flags)) {
        this._values.set(name, this._resolveInitial(def));
      }
    }
  }

  _resolveInitial(def) {
    const init = def.initial;
    if (init === undefined || init === null) {
      return def.type === 'boolean' ? false : '';
    }
    return init;
  }

  get(name) {
    const v = this._values.get(name);
    return v === undefined ? null : v;
  }

  getInt(name) {
    const v = this._values.get(name);
    if (typeof v === 'number') return v | 0;
    return 0;
  }

  getString(name) {
    const v = this._values.get(name);
    return v == null ? '' : String(v);
  }

  getBoolean(name) {
    const v = this._values.get(name);
    return typeof v === 'boolean' ? v : false;
  }

  set(name, value) {
    this._values.set(name, value);
  }

  applyEffect(effect) {
    const val = effect.val;
    switch (effect.op) {
      case 'add': {
        const current = this.getInt(effect.var);
        this._values.set(effect.var, current + (typeof val === 'number' ? val : parseInt(val, 10) || 0));
        break;
      }
      case 'set':
        this._values.set(effect.var, val);
        break;
    }
  }

  applyEffects(effects) {
    if (!effects) return;
    for (const e of effects) this.applyEffect(e);
  }

  snapshot() {
    const obj = {};
    for (const [k, v] of this._values) obj[k] = v;
    return obj;
  }

  toSerializable() {
    return this.snapshot();
  }

  restoreFrom(saved) {
    this._values.clear();
    for (const [k, v] of Object.entries(saved)) {
      this._values.set(k, v);
    }
  }
}
